sql/analyzer/pushdown.go

```diff
@@ -801,139 +803,108 @@ func getIndexesByTable(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scop
-			oi, ok := idx.(sql.OrderedIndex)
-			if !ok || oi.Order() != sql.IndexOrderAsc {
-				continue
+		sfAliases := aliasedExpressionsInNode(s)
+		newN, same, err := transform.NodeWithCtx(s, skipSel, func(tc transform.Context) (sql.Node, transform.TreeIdentity, error) {
+			n := tc.Node
+			// TODO: make this work with IndexedTableAccess, if we are sorting by the same col
+			if ita, ok := n.(*plan.IndexedTableAccess); ok && !ita.IsStatic() {
+				return n, transform.SameTree, nil
 			}
-			if idx.ID() == "PRIMARY" {
-				pkIndex = idx
-				break
+			rs, ok := n.(*plan.ResolvedTable)
+			if !ok {
+				return n, transform.SameTree, nil
 			}
-		}
-		if pkIndex == nil {
-			return s, transform.SameTree, nil
-		}
-
-		// Get primary key column names; these are qualified
-		pkColNames := pkIndex.Expressions()
-
-		// Extract SortField Column Names
-		var sfColNames []string
-		for _, field := range s.SortFields {
-			gf, ok := field.Column.(*expression.GetField)
+			table := rs.Table
+			if w, ok := table.(sql.TableWrapper); ok {
+				table = w.Underlying()
+			}
+			idxTbl, ok := table.(sql.IndexAddressableTable)
 			if !ok {
 				return n, transform.SameTree, nil
 			}
-			// Resolve aliases; aliases should have empty table in GetField
-			if name, ok := aliasMap[gf.String()]; ok {
-				sfColNames = append(sfColNames, name)
-			} else {
-				sfColNames = append(sfColNames, gf.String())
+			idxs, err := idxTbl.GetIndexes(ctx)
+			if err != nil {
+				return nil, transform.SameTree, err
 			}
-		}
 
-		// SortField is definitely not a prefix to PrimaryKey
-		if len(sfColNames) > len(pkColNames) {
-			return s, transform.SameTree, nil
-		}
-
-		// Check if SortField is a prefix to PrimaryKey
-		for i := 0; i < len(sfColNames); i++ {
-			// Stop when column names stop matching
-			if sfColNames[i] != pkColNames[i] {
-				return s, transform.SameTree, nil
+			var pkIndex sql.Index // TODO: support secondary indexes
+			for _, idx := range idxs {
+				if idx.ID() == "PRIMARY" {
+					pkIndex = idx
+					break
+				}
+			}
+			if pkIndex == nil {
+				return n, transform.SameTree, nil
+			}
+			// Some Primary Keys (like doltHistoryTable) are not in order
+			if oi, ok := pkIndex.(sql.OrderedIndex); ok && (!oi.Reversible() || oi.Order() == sql.IndexOrderNone) {
+				return n, transform.SameTree, nil
 			}
-		}
 
-		// Create lookup based off of PrimaryKey
-		indexBuilder := sql.NewIndexBuilder(pkIndex)
-		lookup, err := indexBuilder.Build(ctx)
-		if err != nil {
-			return nil, transform.SameTree, err
-		}
-		if !pkIndex.CanSupport(lookup.Ranges...) {
-			return n, transform.SameTree, nil
-		}
-		newNode, err := plan.NewStaticIndexedAccessForResolvedTable(rs, lookup)
-		if err != nil {
-			return nil, transform.SameTree, err
-		}
+			pkColNames := pkIndex.Expressions()
+			if len(sfExprs) > len(pkColNames) {
+				return n, transform.SameTree, nil
+			}
+			for i, fieldExpr := range sfExprs {
+				if s.SortFields[0].Order != s.SortFields[i].Order {
```

Comment:
This works, we might want to generalize this to more monotonic expressions in the future. Ex, `order by x+1` is OK, `order by 1-x` is OK but is reversed, `order by mod(x)` is not OK. Same monotonic ordering helps with merge joins and stuff.